File: java/src/main/java/io/github/matheuslm7/brvalidator/ie/Go.java
package io.github.matheuslm7.brvalidator.ie;

import io.github.matheuslm7.brvalidator.internal.Shared;
import java.util.regex.Pattern;

/**
 * GO: verified against the official SEFAZ-GO "Roteiro de Crítica da Inscrição Estadual"
 * (sintegra.gov.br/Cad_Estados/cad_GO.html), including its worked example (10.987.654-7).
 * Format: AB.CDE.FGH-I, where AB must be 10, 11, or 20-29.
 */
final class Go {

  private static final Pattern ALLOWED_CHARS = Pattern.compile("^[\\d.\\-\\s]+$");
  private static final int[] WEIGHTS = {9, 8, 7, 6, 5, 4, 3, 2};

  private Go() {}

  static String normalize(String value) {
    return Shared.removeNonDigits(value);
  }

  private static boolean hasValidPrefix(String v) {
    int prefix = Integer.parseInt(v.substring(0, 2));

    return prefix == 10 || prefix == 11 || (prefix >= 20 && prefix <= 29);
  }

  static boolean isValid(String value) {
    if (!ALLOWED_CHARS.matcher(value).matches()) {
      return false;
    }

    String v = normalize(value);

    if (v.length() != 9 || !hasValidPrefix(v)) {
      return false;
    }

    return Mod11.mod11CheckDigit(Mod11.weightedSum(v, WEIGHTS)) == (v.charAt(8) - '0');
  }

  static String format(String value) {
    String v = normalize(value);

    if (v.length() != 9) {
      return value;
    }

    return v.substring(0, 2) + "." + v.substring(2, 5) + "." + v.substring(5, 8) + "-"
        + v.substring(8, 9);
  }
}

File: php/src/Ie/Go.php
<?php

declare(strict_types=1);

namespace BrValidator\Ie;

use BrValidator\Shared\Digits;
use BrValidator\Shared\Mod11;

/**
 * Verified against the official SEFAZ-GO "Roteiro de Critica da Inscricao
 * Estadual" (sintegra.gov.br/Cad_Estados/cad_GO.html), including its
 * worked example (10.987.654-7). Format: AB.CDE.FGH-I, where AB must be
 * 10, 11, or 20-29.
 *
 * Not part of the public API - use BrValidator\Ie.
 */
final class Go
{
    private const ALLOWED_CHARS = '/^[\d.\-\s]+$/';
    private const WEIGHTS = [9, 8, 7, 6, 5, 4, 3, 2];

    private function __construct()
    {
    }

    private static function hasValidPrefix(string $v): bool
    {
        $prefix = (int) substr($v, 0, 2);

        return $prefix === 10 || $prefix === 11 || ($prefix >= 20 && $prefix <= 29);
    }

    public static function normalize(string $value): string
    {
        return Digits::removeNonDigits($value);
    }

    public static function isValid(string $value): bool
    {
        if (!preg_match(self::ALLOWED_CHARS, $value)) {
            return false;
        }

        $v = self::normalize($value);

        if (strlen($v) !== 9 || !self::hasValidPrefix($v)) {
            return false;
        }

        return Mod11::checkDigit(Mod11::weightedSum($v, self::WEIGHTS)) === (int) $v[8];
    }

    public static function format(string $value): string
    {
        $v = self::normalize($value);

        if (strlen($v) !== 9) {
            return $value;
        }

        return substr($v, 0, 2) . '.' . substr($v, 2, 3) . '.' . substr($v, 5, 3) . '-' . substr($v, 8, 1);
    }
}
